src/BookParser.hs

```diff
@@ -436,7 +431,16 @@ pSection = do
   return $ Section header units subsections
 
 pChapter :: (MonadState BookState m, MonadParsec e String m) => (Given ResourcesURI, Given TableOfContents) => m Section
-pChapter = runReaderT pSection (Depth 1) <* eof
+pChapter = do
+  section <- runReaderT pSection (Depth 1) <* eof
+  knownSections <- gets bookStateSectionsDefined
+  usedSections  <- gets bookStateSectionsReferenced
+
+  forM_ (Set.toList usedSections) $ \sectionId ->
```

Comment:
Use set difference, `forM_ (Set.difference usedSections knownSections) $ \sectionId ->`. And you don't `toList` because `Set` is `Foldable` (`forM_` uses any `Foldable`). And finally, I prefer `for_` to `forM_` because its type is less restrictive (you're probably used to `forM_` from Universum, which is the same as `for_`).

Reply:
What about `for_` being inefficient for some data types?

Reply:
This was a bug because these types did not define a faster version of `(*>)` but did define a faster version of `(>>)`. I don't think it applies to `transformers` or `megaparsec` anymore.